frappe/model/document.py

```diff
@@ -89,6 +93,38 @@ def get_doc(*args, **kwargs):
 	raise ImportError(doctype)
 
 
+@contextmanager
+def read_only_document():
+	# Store original methods
+	original_methods = {
+		"save": Document.save,
+		"_save": Document._save,
+		"insert": Document.insert,
+		"delete": Document.delete,
+		"submit": Document.submit,
+		"cancel": Document.cancel,
+		"db_set": Document.db_set,
```

Comment:
Monkey-patching global class methods is an utterly ridiculous change on its own but it also makes the code thread-unsafe.

You went ahead and merged it without any reviews too! https://github.com/frappe/frappe/pull/27955 

@rmehta @akhilnarang @surajshetty3416   I'd seriously suggest revisiting merges without review.


@blaggacao you've also merged several PRs with "chore" title where it's actually modifying production code. What the fuck?

Reply:
Quick off-topic request: could you formulate your reviews without the parts that could end up being received as, unfortunately I have to say, bashing?

I very much appreciate the education part, truly.

But I'm not going to lie, it's also consuming / wasting mental cycles to properly deal with the other parts.

I truly don't need them in order to understand and stand by my responsibility in a professional manner.

The end result may be the same: code is being fixed; code is made better. Hopefully, we can get there as efficiently as possible.

Reply:
On topic: o damn.

Reply:
> "you ship it, you own it; review is sometimes impractical or even impossible"

But you can't own it? Are you doing support tickets at Frappe?

Reply:
> But you can't own it? Are you doing support tickets at Frappe?

My take: "owning" requires awareness, which I'm sourcing from GitHub. In this case, it was sourced about 15min ago and you can assume, as per the precedents, that I'll do something about it within the next 2 days.

That's as much ownership as I can technically contribute.

Reply:
> you should at least get approval for critical PRs.

Yes, the crux lies in the classification. 

For example, this PR isn't technically critical, because it doesn't do anything on its own. It's simply a feature not yet stabilized.

If anything, the critical PR rather was: https://github.com/frappe/frappe/pull/27956

A "feature not yet stabilized" could well have a very different treatment if we had the semantics for it.